design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,217 @@
+desired settings. The BIOS Registry can be used as a source of documentation
+to modify the settings.
+
+## Design Details
+
+This proposes three new data elements in the BareMetalHost - the current BIOS
+settings retrieved from Ironic, the BIOSRegistry read from Ironic (effectively
+a schema for the BIOS settings), and a bulk BIOS settings for Ironic to write
+to the BMC. The current BIOS settings and BIOS Registry can be used to create
+the bulk BIOS settings which can be set on multiple hosts of the same type,
+the current settings and registry will not be used on these hosts.
+
+The current BIOS settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`. These BIOS settings
+can be read via the Ironic API (`/v1/nodes/{node_ident}/bios`) as a set of
+name:value pairs. Ironic can set the BIOS Configuration values through the use
+of clean_steps as part of the manual cleaning process as described in the Ironic
+documentation. BIOS settings are configured as name/value pairs and can include
+from one to multiple pairs.
+
+What we are proposing is to make the cached BIOS settings from Ironic available
+to the user. This can be either via a ConfigMap or special purpose CRD
+referenced in the BareMetalHostStatus.  As there are no fields in BIOS
+configuration that are secret (i.e. no passwords etc.) this new entry does not
+need to be secret-protected. 
+
+A new Ironic API must be provided for the attributes section of the BIOS
+registry (`/v1/nodes/{node_ident}/bios_registry`) as this is currently not
+supported in Ironic. The registry contains a JSON schema describing the BIOS
+settings - names, types, enumeration values, readonly flag etc. The BIOS
+registry will be retrieved from the BMC by Ironic at the same time Ironic
+gets the BIOS settings, as described in the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681).
+An example of an attributes entry in the BIOS registry is:
+
+    {
+        "AttributeName": "SriovGlobalEnable",
+        "CurrentValue": null,
+        "DisplayName": "SR-IOV Global Enable",
+        "DisplayOrder": 2331,
+        "HelpText": "Enables or disables the BIOS configuration of Single Root
+                     I/O Virtualization (SR-IOV) devices. Enable this feature
+                     if booting to a virtualization operating system that
+                     recognize SR-IOV devices. ",
+        "Hidden": false,
+        "Immutable": false,
+        "MenuPath": "./IntegratedDevicesRef",
+        "ReadOnly": false,
+        "ResetRequired": true,
+        "Type": "Enumeration",
+        "Value": [
+          {
+            "ValueDisplayName": "Enabled",
+            "ValueName": "Enabled"
+          },
+          {
+            "ValueDisplayName": "Disabled",
+            "ValueName": "Disabled"
+          }
+        ],
+        "WarningText": null,
+        "WriteOnly": false
+    }
```

Comment:
How firm is this part of the standard? If most vendors support s similar schema, it seems like we could expose a subset of the information in a CRD using a similar format. The `Name`, `CurrentValue`, `Type`, and `ReadOnly` fields look especially useful. Maybe the `Value`, too, since that seems to be related to the enum?

Do integer values have min and max settings?

Reply:
Interesting that they have an OpenAPI specification for their custom non-OpenAPI schema :D
I wonder how hard it would be to slim this down and transform the format itself into OpenAPI, for consistency with how k8s APIs are specified, since afaik it's not actually our goal for someone to be able to build a GUI on top of this.